libs/model/src/user/Xp.projection.ts

```diff
@@ -0,0 +1,193 @@
+import { Projection, events } from '@hicommonwealth/core';
+import * as schemas from '@hicommonwealth/schemas';
+import {
+  QuestParticipationLimit,
+  QuestParticipationPeriod,
+} from '@hicommonwealth/schemas';
+import { isWithinPeriod } from '@hicommonwealth/shared';
+import { Op } from 'sequelize';
+import { z } from 'zod';
+import { models, sequelize } from '../database';
+import { mustExist } from '../middleware/guards';
+
+const inputs = {
+  ThreadCreated: events.ThreadCreated,
+  CommentCreated: events.CommentCreated,
+  CommentUpvoted: events.CommentUpvoted,
+  //PollCreated: events.PollCreated,
+  //ThreadEdited: events.ThreadEdited,
+  //CommentEdited: events.CommentEdited,
+  //PollEdited: events.PollEdited,
+};
+
+async function getUserId(payload: { address_id: number }) {
+  const address = await models.Address.findOne({
+    where: { id: payload.address_id },
+    attributes: ['user_id'],
+  });
+  mustExist('Address not found', address);
+  return address.user_id!;
+}
+
+async function getQuestActionMeta(
+  event_payload: { community_id: string; created_at?: Date },
+  event_name: keyof typeof events,
+) {
+  // make sure quest was active when event was created
+  const quest = await models.Quest.findOne({
+    where: {
+      community_id: event_payload.community_id,
+      start_date: { [Op.lte]: event_payload.created_at },
+      end_date: { [Op.gte]: event_payload.created_at },
+    },
+    include: [
+      { required: true, model: models.QuestActionMeta, as: 'action_metas' },
+    ],
+  });
+  if (!quest || !quest.action_metas) return;
+
+  const action_metas = quest.get({ plain: true }).action_metas!;
+  const action_meta = action_metas.find((a) => a.event_name === event_name);
+  if (action_meta)
+    return {
+      ...action_meta,
+      event_created_at: event_payload.created_at!,
+    };
+}
+
+async function recordXps(
+  user_id: number,
+  action_meta: z.infer<typeof schemas.QuestActionMeta> & {
+    event_created_at: Date;
+  },
+  creator_user_id?: number,
+) {
+  const event_created_at = action_meta.event_created_at;
```

Comment:
If we only have an "at least once" delivery guarantee, won't this break in the event of a double event process because it will record the action twice?

Reply:
yes, it should be idempotent... let's try find or create